design/lakectl-testing.md

```diff
@@ -0,0 +1,102 @@
+  * It is obvious that changing the test **format** (e.g. expanding) will require framework modifications, in order to support the modified format
+* Decoupling of Framework and Data - the test data format should contain all the required info for the test itself, regardless of the framework design. In other words, we would like the ability to rewrite/replace the framework and still use the same tests data
+*
+
+## What We Want To Test 
+
+The goal is to the test `lakectl` correctness. At the very least we would like to verify that:
+* CLI input is parsed correctly
+* The correct requests are sent to the server (if any)
+* Server result is formatted correctly
+* Both erroneous input and error results from the server are formatted correctly
+* CLI Internal errors are handled correctly
+
+This testing framework is not designed to perform end-to-end tests (i.e. server correctness) or performance, for that matter
+
+## Architecture
+
+### Architecture Outline
+
+![lakectl testing architecture](./diagrams/lakectl-testing-arch.png)
+
+### Test Case Structure
+
+The test case structure should contain at least the following details:
+* pre-condition: the system state right before the tested command starts. This can be a function name to call, a mock type to use or a file containing lakectl commands to run in order to bring the system to that state (in case real server is in use)
+* lakectl command: a string with the `lakectl` command to test.
+* expected result: the expected output of a correct execution of the above command. This can be a string or, better, a file containing the output string
+
+In the future we might consider expending this structure with more fields, to allow extended functionality
+ 
+### How should we use `lakefs` backend
+
+Here are some of the pros and cons of each approach
+
+#### **Live Server**
+
+On the upside:
+* Straightforward - the implementation is already there, just bring the server up.
+  * Moreover, `lakefs` modifications are immediately reflected into the tests
+* The best possible simulation - this is the real server `lakectl` is communicating with
+
+And then, on the downside:
+* Bringing up the server requires additional dependencies - underlying storage, DB etc. While this supports the advantage of testing a **real** server environment, it complicates the test setup, and although this is easily solved by containers, it still worth mentioning
+* Performance-wise, creating a server environment for each test (thousands of them, ideally) is very expensive. We can, of course, group several tests (all of them?) to use a single server session, but this will eliminate the isolative and stateless nature of the tests, which is undesired
+* Flexibility is limited to `lakefs` server implementation. This means that testing `lakectl` behavior, with improbable response from `lakefs` server (e.g. timeout, disconnection etc.) becomes very difficult
```

Comment:
Commented by @itaiad200 
Kind of means that this test framework also becomes sort of a system test for lakeFS, which I believe we should be ok with.

Reply:
In that case I think it is no longer `lakectl` testing, but system/end2end testing, which, BTW I'm totally OK about - but then we shoudl change the scope of this issue and design.
Maybe that's what we should do?!

Reply:
lakeFS system testing is covered by Nessie. I think that we should scope this to `lakectl` testing, which is easier and more reliable when you have a running lakeFS. I mentioned that lakeFS is being tested too, because now, for some system flows, will be covering it more than once. There's no downside to it IMO, as the focus of this tests is `lakectl`.